omisego-sdk/src/main/java/co/omisego/omisego/OMGAPIClient.kt

```diff
@@ -0,0 +1,96 @@
+package co.omisego.omisego
+
+/**
+ * OmiseGO
+ *
+ * Created by Phuchit Sirimongkolsathien on 5/3/2018 AD.
+ * Copyright © 2017-2018 OmiseGO. All rights reserved.
+ */
+
+import co.omisego.omisego.custom.Callback
+import co.omisego.omisego.custom.CallbackManager
+import co.omisego.omisego.custom.Serializer
+import co.omisego.omisego.model.*
+import co.omisego.omisego.network.ewallet.EWalletClient
+import com.google.gson.reflect.TypeToken
+
+/**
+ * The class OMGApiClient represents an object that knows how to interact with OmiseGO API.
+ *
+ * Create instances using [OMGApiClient.Builder] and pass your implementation of [Callback<T>] interface
+ * to generate an implementation
+ *
+ * For example,
+ * <code>
+ * val omgApiClient = OMGApiClient.Builder {
+ *      setAuthorizationToken(YOUR_TOKEN)
+ * }.build()
+ *
+ * omgApiClient.getCurrentUser(object : Callback<User> {
+ *      override fun success(response: OMGResponse<User>) {
+ *          // Do something
+ *      }
+ *
+ *      override fun fail(response: OMGResponse<ApiError>) {
+ *          // Handle fail case properly
+ *      }
+ * })
+ * </code>
+ *
+ */
+class OMGAPIClient(private val eWalletClient: EWalletClient) {
+    private val serializer = Serializer()
+
+    /**
+     * Asynchronously send the request to transform the [User] corresponding to the provided authentication token.
+     * if *success* the [callback] will be invoked with the [User] parameter,
+     * if *fail* [callback] will be invoked with the [co.omisego.omisego.models.ApiError] parameter.
+     *
+     * @param callback A callback to receive the response from server.
+     */
+    fun getCurrentUser(callback: Callback<User>) {
+        val type = object : TypeToken<OMGResponse<User>>() {}.type
+        val callbackManager = CallbackManager<User>(serializer, type)
```

Comment:
I've tried to use a combo of `inline` and `reified` type. Unfortunately, I don't know why the `Gson` cannot convert the response to the specific model, but instead `LinkedTreeMap`. The weird thing is that when I ran it on the android device, it just throw that exception, but for junit test it doesn't happen 😂

Reply:
I've tried your solution and it failed the JUnit

```
Argument(s) are different! Wanted:
callback.success(
    OMGResponse(version=1, success=true, data=User(id=48236187-9c5c-4568-8cdf-f0233d035574, providerUserId=provider_user_id01, username=user01, metaData=null))
);
-> at co.omisego.omisego.OMGAPIClientTest.OMGAPIClient should call get_current_user and success(OMGAPIClientTest.kt:108)
Actual invocation has different arguments:
callback.success(
    OMGResponse(version=1, success=true, data={username=user01, provider_user_id=provider_user_id01, object=user, metadata={}, id=48236187-9c5c-4568-8cdf-f0233d035574})
);
```

Reply:
The exception for the JUnit test seems to be a different issue: you're expecting `metaData` to be null but it's empty in the actual result. Did you update to use `emptyMap()` by default instead of null? 🤔